### DISCUSSION TITLE
[Share] Attachment Rename

### DISCUSSION THREAD

**Discussion by thalient-ai** [windingwind/zotero-actions-tags#380] [Action Scripts]
### Description

In Zotero, you can rename attachment filenames by selecting the attachments (not the parent items) right-clicking and selecting 'Rename Files from Parent Metadata'. Selecting this option changes the attachment filename but not the attachment title.

This script lets you:
- Rename both the attachment title and the filename
- Execute the script on parent times (one or many)

It is simply a more user-friendly option than trying to select multiple attachments to rename.

### Event

None

### Operation

Script

### Data

```javascript
/**
 * Bulk rename of attachment titles
 * @author thalient-ai
 * @usage 
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/380
 * @see https://github.com/windingwind/zotero-actions-tags/discussions/380
 */

const Zotero = require("Zotero");

// Function to process renaming of each attachment
async function processRenaming(attachment, processedAttachmentIds) {
    if (!attachment || processedAttachmentIds.has(attachment.id)) {
        // Skip processing if attachment is undefined or already processed
        return { renamed: 0, errors: 0 };
    }

    if (attachment.attachmentLinkMode === Zotero.Attachments.LINK_MODE_LINKED_URL) {
        Zotero.logError(`Cannot rename linked URL attachment ${attachment.id}.`);
        return { renamed: 0, errors: 1 };
    }

    if (!attachment.parentItemID) {
        Zotero.logError(`Attachment ${attachment.id} does not have a parent item.`);
        return { renamed: 0, errors: 1 };
    }

    const parentItem = await Zotero.Items.getAsync(attachment.parentItemID);
    if (!parentItem) {
        Zotero.logError(`No parent item found for attachment ${attachment.id}.`);
        return { renamed: 0, errors: 1 };
    }

    const currentPath = await attachment.getFilePathAsync();
    if (!currentPath) {
        Zotero.logError(`No local file path available for attachment ${attachment.id}.`);
        return { renamed: 0, errors: 1 };
    }

    // Updated to include attachmentTitle
    const newName = Zotero.Attachments.getFileBaseNameFromItem(parentItem, { attachmentTitle: attachment.getField("title") });
    const currentName = currentPath.split(/(\\|\/)/g).pop();
    const extension = currentName.includes('.') ? currentName.split('.').pop() : '';
    const finalName = extension ? `${newName}.${extension}` : newName;

    if (newName !== currentName) {
        try {
            await attachment.renameAttachmentFile(finalName);
            // Set the title without the file extension
            attachment.setField('title', newName);
            await attachment.saveTx();
            processedAttachmentIds.add(attachment.id);
            return { renamed: 1, errors: 0 };
        } catch (error) {
            Zotero.logError(`Error renaming attachment ${attachment.id}: ${error}`);
            return { renamed: 0, errors: 1 };
        }
    }

    return { renamed: 0, errors: 0 };
}

// Main execution block
(async () => {
    if (!items && !item) {
        Zotero.alert(null, "Rename Attachments", "[Rename Attachments] No item or items array provided.");
        return;
    }

    // Simplified targetItems logic
    let targetItems = items || [item];
    let totalRenamed = 0;
    let totalErrors = 0;
    let processedAttachmentIds = new Set();

    for (const currentItem of targetItems) {
        const attachments = currentItem.itemType === 'attachment' ? [currentItem] : await Zotero.Items.getAsync(currentItem.getAttachments());
        for (const attachment of attachments) {
            const result = await processRenaming(attachment, processedAttachmentIds);
            totalRenamed += result.renamed;
            totalErrors += result.errors;
        }
    }

    // Display a summary alert only if there are significant outcomes to report
    if (totalRenamed > 0 || totalErrors > 0) {
        Zotero.alert(null, "Rename Attachments", `[Rename Attachments] Successfully renamed ${totalRenamed} attachments. Errors: ${totalErrors}`);
    }
})();

```

### Anything else

The script has error handling for linked attachments, attachments that are not local, and attachments without parents.

Attachment filenames will include the file extension (e.g., .pdf)
Attachment titles will not include the file extension

**Comment by Deg2v** [comment]
This is a very good script, but I found that it will rename all attachments under the entry. How can I rename only the body pdf (i.e. the pdf opened by double-clicking the entry)?

**Comment by crnkv** [comment]
According to the [Plugin README](https://github.com/windingwind/zotero-actions-tags/blob/master/README.md#custom-script), the script will be called twice: one is for all items (`items=[...], item=undefined`), the other is for each item individually (`items=[], item=...`).
Either case, the line in your script `let targetItems = items || (item ? [item] : [])` will always be equivalent to `let targetItems = items` because `[]` (empty array) is true.
Maybe you want to avoid the case `items=null` when the script is triggered in a way other than menu/shortcut (see the Plugin README), but this scenario also doesn't quite seem to be possible...
Plus, since the case `(!items && !item)` has already been ruled out, so even if the case `items=null` happens to exist, then `item` would be automatically true. It would be logically complete to write `let targetItems = items || [item]`.

On the contrary, the line `const newName = Zotero.Attachments.getFileBaseNameFromItem(parentItem)` is lacking an important feature supported by Zotero's Filename Format Template -- the `attachmentTitle` variable.
It should be `const newName = Zotero.Attachments.getFileBaseNameFromItem(parentItem, {attachmentTitle: attachment.getField("title")})` to implement this.

**Reply by thalient-ai** [reply]
Updated, thank you for the feedback
